### PR TITLE
fix(guardrails): stop the here-string deadlock that voided two blocking guards

### PR DESCRIPTION
**Pull request by kyle-sexton**
No linked issue

## Summary

Two BLOCKING PreToolUse guards — `secret-pattern-detection` and `hardcoded-path-check` — returned
**no verdict at all** for a Write/Edit payload of **65536–65663 bytes inclusive**. Not slow:
deadlocked. A live-shape AWS access-key id inside such a payload produced nothing; the same token in
a small payload exits 2. Both hooks are registered at `timeout: 60`, so the harness cancels the guard
and the verdict is lost — a fail-open reachable by any agent that controls the size of what it writes.

Same class as #1587, which fixed `hook-utils.sh`'s JSON path and stopped at that one call site. This
PR sweeps the class instead of patching only the two reported files.

## The defect

Bash delivers a here-string by filling a pipe **itself**, before the reader is `exec`'d, and it
appends a newline. A payload in that band puts the write 1–128 bytes past the 65536-byte pipe
capacity and bash blocks forever. At ≥129 bytes over, bash spills to a temp file and it works again —
so the window is **closed on both sides**, which is exactly why no ordinary size ever caught it.

Reproduced standalone, outside the plugin (`bash 5.3.15(1)`, MINGW64):

```
C=$(head -c 65600 /dev/zero | tr '\0' b)
timeout 15 bash -c 'grep -qE "Users" <<<"$1"' _ "$C"; echo $?   # 124 (hung)
# 65535 -> 1   65536 -> 124   65600 -> 124   65663 -> 124   65664 -> 1
```

The `while … done <<<"$var"` shape hangs identically (rc 124 at 65536 / 65600 / 65663), which is what
pulled the command-scanning guards into scope.

## Boundary measurements, through the real hooks

Payload piped to the hook on stdin — never `<<<`, which would hang the measurement itself. `rc 124` =
killed at the bound, i.e. never answered. All numbers from this Windows host (Git Bash + Defender),
under concurrent agent load.

`secret-pattern-detection`, exact content bytes, AWS access-key id at the tail:

| content bytes | case | BEFORE rc | BEFORE secs | AFTER rc | AFTER secs |
| --- | --- | --- | --- | --- | --- |
| 65535 | clean | 0 | 10 | 0 | 19 |
| 65535 | AWS key | 2 | 43 | 2 | 63 |
| 65536 | clean | **124** | 202 (bound) | **0** | 14 |
| 65536 | AWS key | **124** | 204 (bound) | **2** | 51 |
| 65600 | clean | **124** | 151 (bound) | **0** | 12 |
| 65600 | AWS key | **124** | 151 (bound) | **2** | 67 |
| 65663 | clean | **124** | 156 (bound) | **0** | 24 |
| 65663 | AWS key | **124** | 155 (bound) | **2** | 41 |
| 65664 | clean | 0 | 20 | 0 | 19 |
| 65664 | AWS key | 2 | 44 | 2 | 59 |

The BEFORE hangs were taken at a **200-second** bound first, then 150 — well past the legitimate slow
path (41–67 s) — so these are deadlocks, not slowness.

`hardcoded-path-check`, clean payloads. The pre-filter gate runs on **every** write, so the clean
column is the stronger claim: nothing in the window got a verdict, violating or not.

| content bytes | BEFORE rc | BEFORE secs | AFTER rc |
| --- | --- | --- | --- |
| 65535 | 0 | 34 | 0 |
| 65536 | **124** | 152 (bound) | **0** |
| 65600 | **124** | 151 (bound) | **0** |
| 65663 | **124** | 207 (bound) | **0** |
| 65664 | 0 | 44 | 0 |

## Why not `printf … | grep -q` — the pipefail inversion is real

`hardcoded-path-patterns.sh:73-75` carried a comment *justifying* the here-string, and the
justification was half right. Measured under `set -o pipefail`, with the match on line 1 so `grep -q`
can exit before the writer finishes (a single-line payload never reproduces this — grep must read it
all, so `printf` never gets SIGPIPE'd):

| shape | `set +o pipefail` | `set -o pipefail` |
| --- | --- | --- |
| `grep -qE pat <<<"$C"` | 0 | 0 — but deadlocks in the window |
| `printf … \| grep -qE pat` | 0 | **141** ← inversion |
| `printf … \| grep -E pat >/dev/null` | 0 | 0 |
| `grep -qE pat < <(printf …)` | 0 | **0** ← chosen |

Both hooks run under `set -uo pipefail` (`set -e` is off), so the inversion is live — and it is worse
than a wrong status, because both gate sites are written `if ! grep -q …`:

> `grep -q` matches → exits 0 → SIGPIPEs `printf` → `pipefail` reports **141** → `if ! 141` is
> **true** → the gate early-returns **clean**. A fail-open on the very payload that contained the
> secret.

**Chosen idiom: process substitution.** It keeps the writer *outside* the pipeline, so `pipefail` can
never see its SIGPIPE, while preserving the `-q` early exit the gate exists for — and it never
blocks. Verified at all six sizes in both the match and no-match directions.

Two shapes, chosen by whether the reader drains its input. The pattern lib previously
**contradicted `hook-utils.sh` inside the same plugin** — it told readers to PREFER a here-string
over `printf | grep`, while `hook-utils.sh` told them a whole payload must never go through `<<<`.
The lib now states the same rule and cites it:

- reader drains (`jq`, `grep` without `-q`) → `printf … | reader`
- reader may exit early (`grep -q`) → `reader < <(printf …)`

For `while … done` loops the substitution is `< <(printf '%s\n' …)`. The `\n` is mandatory and makes
it byte-identical to the here-string it replaces (`<<<` appends a newline unconditionally), so no
loop can drop its final line.

## Repo-wide sweep of `<<<` — every site, with a verdict

876 occurrences total; 329 outside `*.test.sh`. Fix criterion: **the string can reach 65536–65663
bytes from an agent- or attacker-controlled source, AND a hang loses a security verdict.**

### Fixed (18 sites)

| site | input | why |
| --- | --- | --- |
| `guardrails/lib/path-detection/hardcoded-path-patterns.sh:76,83` | whole Write/Edit payload | reported; blocking |
| `guardrails/hooks/secret-pattern-detection.sh:158,207` | whole Write/Edit payload | reported; blocking |
| `guardrails/hooks/hardcoded-path-check.sh:219` | `$VIOLATIONS` | **not in the original report.** `$VIOLATIONS` embeds each MATCHED LINE verbatim, and the lib's `head -3` bounds the line COUNT, not bytes — so one 65KB minified line carrying a hardcoded path makes it payload-sized. It deadlocks on the **blocked** path, after the stderr message but before `exit 2`. Measured separately below |
| `guardrails/hooks/block-convention-violation.sh:132,158` | `$cmd` (Bash/PowerShell command) | blocking guard; loop shape hangs identically |
| `guardrails/hooks/block-hook-bypass.sh:248,497,566` | `$cmd`, `$NORMALIZED_SEGMENTS` (derived from `$COMMAND`) | blocking guard |
| `guardrails/hooks/flag-commit-pr-skill-bypass.sh:229` | `$cmd` | same command stripper |
| `guardrails/lib/powershell/ps-command.sh:143,671` | `$cmd`, `$norm` | shared lib behind the blocking PowerShell guards |
| `guardrails/hooks/workflow-resilience-check.sh:72,79` | `$SCRIPT` (inline Workflow script, or a `scriptPath` file read) | advisory, so no verdict is lost — but a hang wedges the Workflow call until the harness cancels |
| `source-control/hooks/pr-body-linkage-gate.sh:194` | `$text` (PR body) | **blocking** gate. GitHub caps a PR body at exactly **65536 characters** — the documented maximum lands inside the hang window |
| `source-control/hooks/pr-linkage-validator.sh:76,113` | `$body` | same input, same cap |

`source-control` is deliberately in scope rather than left as a half-fix; it costs the second plugin
bump in this PR.

### Judged safe — no fix, with reason

| site | reason |
| --- | --- |
| `guardrails/lib/verification/verify-cli-flag.sh:159,161` (`$HELP_OUTPUT`) | local CLI `--help` output; not attacker-influenced. A **latent hang**, not a security hole — noted, not fixed |
| `guardrails/hooks/cli-flag-verify.sh:188,337` (`read -ra` on `$seg`/`$chainstr`) | advisory PostToolUse; a single ≥64KB fragment is possible in principle, but no verdict is at stake. Latent hang, noted |
| `guardrails/hooks/flag-commit-pr-skill-bypass.sh:196` (`$keys`), `skill-reference-verify.sh:160` (`$declared`) | jq-derived plugin/settings key names; bounded by manifest size, not by any payload |
| `guardrails/hooks/block-no-verify.sh:102` | a `userConfig` option value (administrator-provided scalar) |
| `biome-format:211,237,251`, `ruff-format:250,275` (`$OUTPUT`) | formatter/linter output. Same mechanism, **different consequence class** — no security verdict at stake, only a wedged formatter. Recommended follow-up, kept out to keep this PR reviewable |
| `claude-ops-paths.sh:19,68`, `worktree-create.sh:373,412`, `babysit-readiness-gate.sh:286,293`, `check-plugin-manifest-presence.sh:75` | `IFS=… read -ra` splits of a path or a short CSV; cannot approach 64KB |
| `source-control/skills/pull-request/scripts/fetch-annotations.sh:177` (`$FILTERED`) | jq-filtered CI annotation records in a skill script, not a hook gate; no blocking verdict |
| `claude-config/skills/audit/scripts/*`, `claude-ops/skills/lanes/scripts/*`, `work-items` adapters | skill-invoked scripts over jq-bounded JSON, not a hook payload; no blocking verdict |
| ~547 sites in `*.test.sh` | fixed small fixtures authored in-repo; reported as one class |

`lib/hook-utils.sh` itself was already clean (#1587), and the stdin→`CONTENT` path in both hooks is
`printf '%s' "$INPUT" | jq -r` throughout — verified, because otherwise the payload would have hung
upstream and this fix would have changed nothing.

### The `$VIOLATIONS` site, measured

A BEFORE run cannot reach line 219 through the hook (the gate deadlocks first), so it is measured
directly, and its reachability is confirmed on the patched hook:

```text
$VIOLATIONS as hpp::scan_text builds it: the label, then "<lineno>:<line>" with the
matched line embedded VERBATIM, then the block terminator.  bytes = 65628

PRE-FIX   grep -E 'detected:$' <<<"$VIOLATIONS"    -> rc=124  (deadlock, 60s bound)
POST-FIX  grep -E 'detected:$' < <(printf '%s' …)  -> rc=0    (4s under load)
```

## Tests

Neither suite had a **single** payload-size case before this (`grep -n '65536\|head -c'` returned
nothing). Added to **both** `secret-pattern-detection.test.sh` and `hardcoded-path-check.test.sh`:

- clean payloads at **65535, 65536, 65600, 65663, 65664** — the window and both shoulders;
- a **real detectable secret / hardcoded path inside the window** (65536 and 65600) that must exit 2;
- the empty-content case, pinning that `printf '%s' ""` (zero bytes) matches the old `<<<""` (one
  empty line) in outcome;
- a stderr assertion that the `grep -q` early-exit leaks no `Broken pipe` noise onto the hook's
  user-facing channel.

Every case is bounded by `timeout 150` so a regression **fails loudly instead of hanging CI**, and
asserts the **exact** expected code — `124` is reported as its own named failure. A "non-zero means
blocked" assertion would have accepted the hang and would not have caught this defect. The payload is
piped, never fed to the hook with `<<<`, which would hang the test itself at exactly these sizes.

In the path suite the detect payload separates the filler from the home path with a **space**: the
slash-rooted macOS/Linux bodies require a left boundary, so a path glued straight onto filler bytes
legitimately does not match — and the "must block" case would have passed for the wrong reason. That
was caught by a measurement script that omitted the space and returned 0 where 2 was expected.

## Known follow-up (not fixed here)

On this Windows host the **patched** detect path measured **41–67 s** against the `timeout: 60`
registration in `hooks.json`. So on Git Bash under Defender a large payload can still lose its verdict
to the harness — now by slowness rather than deadlock. The cost is process spawns, not matching: on a
hit, itemization runs 12 patterns × 5 processes. That is a separate defect with a separate fix
(batch the itemization), deliberately out of scope here, and stated rather than left for the next
auditor to "discover" as a half-fix.

## Verification

- `shellcheck -x` clean on all 17 changed files.
- `shfmt -d` clean on every changed file. The pre-existing drift in the two `.test.sh` files is
  unchanged by this PR — confirmed byte-identical at `origin/main` — and does not touch the added block.
- `scripts/check-shell-portability.sh --paths …` — no unexcused GNU-only constructs in 27 shell files.
- `scripts/check-changelog-parity.sh --check-bump origin/main` — passes.
- `scripts/sync-hook-utils.sh --check` and `--check-bump origin/main` — pass.

`lib/hook-utils.sh` is left **byte-identical to `main`** on purpose. Its guidance was already correct
(only its upper bound was imprecise), and the sync gate requires a version bump plus a changelog entry
for **all fourteen** other plugins carrying the shared lib in exchange for a comment-only edit — churn
that would bury a security fix. The contradiction is resolved on the guardrails side, which is where
the wrong advice lived.
- `markdownlint-cli2` — 0 issues.
- `secret-pattern-detection.test.sh` — **PASS=52 FAIL=0**
- `hardcoded-path-check.test.sh` — **PASS=94 FAIL=0**

**Version bumps are patch, deliberately.** A reviewer may reach for minor on "payloads that were
allowed are now blocked". Nothing legitimate becomes refused that the guard did not already intend to
refuse — the fix restores the documented contract rather than widening it, which matches this repo's
practice (`source-control` 0.49.3 shipped a behavior-changing `exec-bit-check` fix at patch level;
the `guardrails` 0.21.0 minor was called out specifically for an *acceptance* change that could
refuse previously-allowed legitimate work).

## Rider

The README hook table listed all six guards registered under the `Bash|PowerShell` matcher as
`PreToolUse · Bash`; no row named PowerShell at all. Verified row-by-row against `hooks.json`
(6 rows, 6 hooks, exact match) and corrected.

## Related

- Refs #1587 — fixed this same deadlock class in `hook-utils.sh`'s JSON path and stopped at that one
  call site; this PR sweeps the rest.
- Refs #2007 — prior `guardrails` fix in the same review stream.

### CODE REVIEW

**Comment by kyle-sexton** [comment]
## Suite evidence

**CI is authoritative and covers every affected suite.** `plugin-gate` runs
`scripts/run-plugin-tests.sh`, which discovers *every* `plugins/**/*.test.sh` plus
`.claude/hooks/*.test.sh` — so all eight suites covering files this branch touches ran there, on
Linux, including the new boundary cases.

**All 33 checks pass**, including the three required ones:

| check | result |
| --- | --- |
| `pr-title` | pass |
| `do-not-merge` | pass |
| `ci-status` | pass |
| `plugin-gate` (every `plugins/**/*.test.sh`) | pass, 5m19s |
| `hook-utils-sync` | pass |
| `hook-utils-windows` | pass |
| `changelog-parity-gate` | pass |
| `pr-issue-linkage` | pass |
| `shell-portability-lint`, `portability-lint`, `silent-skip-gate`, `cross-plugin-source-drift` | pass |

`hook-utils-sync` failed on the first push and drove a deliberate scope decision: the gate requires
**every** plugin carrying the shared lib to bump when it changes, so a comment-only edit to
`lib/hook-utils.sh` would have cost fourteen extra plugin bumps and fourteen changelog entries. The
lib is now byte-identical to `main` and the reconciliation lives on the guardrails side, where the
incorrect advice actually was. See the third commit.

### Local runs (Git Bash on Windows — the host where the deadlock reproduces)

Recorded because CI runs Linux, and the defect was found on Windows:

| suite | result |
| --- | --- |
| `secret-pattern-detection.test.sh` | **PASS=52 FAIL=0** (10 new boundary assertions) |
| `hardcoded-path-check.test.sh` | **PASS=94 FAIL=0** (10 new boundary assertions) |
| `block-convention-violation.test.sh` | **PASS=31 FAIL=0** |
| `block-hook-bypass.test.sh` | 50 assertions, 0 failures at the point the local run was stopped as redundant with `plugin-gate` |

All 20 new boundary assertions pass on both hosts:

```text
ok: boundary: clean 65535-byte payload → exit 0
ok: boundary: clean 65536-byte payload → exit 0
ok: boundary: clean 65600-byte payload → exit 0
ok: boundary: clean 65663-byte payload → exit 0
ok: boundary: clean 65664-byte payload → exit 0
ok: boundary: AWS key in 65536-byte payload → exit 2      (path suite: linux home path)
ok: boundary: AWS key in 65600-byte payload → exit 2      (path suite: linux home path)
ok: boundary: in-window block still reports the label
ok: boundary: no SIGPIPE noise on stderr
ok: boundary: empty content → exit 0
```

Every one of these is bounded by `timeout 150` and asserts an exact exit code, with `124` reported as
its own named failure — so a reintroduced deadlock fails loudly rather than hanging the job.

File: plugins/guardrails/hooks/secret-pattern-detection.sh [hunk 1]
```diff
@@ -153,9 +153,18 @@ emit_tel() {
...
 LABELS=()
 
+# Feeds $CONTENT through PROCESS SUBSTITUTION, never `<<<"$CONTENT"`. Bash
+# delivers a here-string through a pipe and appends a newline, so a payload of
+# 65536-65663 bytes puts the write 1-128 bytes past the 65536-byte pipe capacity
+# and bash blocks FOREVER (at >=129 bytes over it spills to a temp file and
+# works again). This guard blocks, so a hang means the harness cancels it at the
+# hook timeout and the secret verdict is lost entirely. `printf … | grep` is not
+# the fix either — see the full rule at the gate in
+# lib/path-detection/hardcoded-path-patterns.sh. `printf '%s'` matches the
```
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply plugins/guardrails/hooks/secret-pattern-detection.sh]: Verified and fixed in 2895890c — you're right, and the correction is worth stating precisely.

**Your premise checks out.** `check_pattern`'s reader takes no `-q`, drains its input, and its pipeline status is discarded (only the captured `$lines` is read). A plain `printf … | grep` would have been correct at this site.

**But the conversion was still required here**, for a reason the old comment buried: the deadlock is a property of `<<<` alone, *not* of the reader. Measured at 65600 bytes on Git Bash:

```text
grep -nE "Users" <<<"$C"                 -> rc=124  (47s bound — deadlocked)
printf '%s' "$C" | grep -nE "Users"      -> rc=1    (4s)
grep -nE "Users" < <(printf '%s' "$C")   -> rc=1    (4s)
```

That first line is this exact non-`-q`, fully-draining reader hanging on a here-string just as `grep -q` does. So `<<<` had to go regardless; only the *choice* of replacement depends on whether the reader early-exits.

I also checked the real pipeline, since `head -3` truncates and can SIGPIPE `grep` transitively — both shapes behave identically under `pipefail`:

```text
printf '%s' "$D" | grep -nE "Users" | head -3 | cut -d: -f1    -> rc=0 lines=[1]
grep -nE "Users" < <(printf '%s' "$D") | head -3 | cut -d: -f1 -> rc=0 lines=[1]
```

**Action taken:** loosened the comment rather than changing working code in a security fix. It now states that the deadlock is reader-independent (with the measurement), that a plain pipe would also have been correct here, and that process substitution is used for **uniformity** with the `-q` gates in this plugin — where a pipe genuinely *would* invert the verdict under `pipefail` — so all the scan sites read as one idiom.

One correction to the thread: `hardcoded-path-check.sh:226` never carried the misleading sentence. Its comment only says "Process substitution, not `<<<`" and then explains why `$VIOLATIONS` is payload-sized, so it needed no change.

File: plugins/source-control/hooks/pr-linkage-validator.sh [hunk 2]
```diff
@@ -110,7 +110,7 @@ has_linkage() {
 related_section() {
   local body="$1" line t start=0 lvl i=0 out=""
   local -a lines=()
-  while IFS= read -r line || [[ -n "$line" ]]; do lines+=("$line"); done <<<"$body"
+  while IFS= read -r line || [[ -n "$line" ]]; do lines+=("$line"); done < <(printf '%s\n' "$body")
```
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply plugins/source-control/hooks/pr-linkage-validator.sh:114]: Verified and fixed in 2895890c.

Confirmed against the source: line 113 was indeed the only converted site in this PR without the trailing note, while `strip_html_comments` 37 lines above it carries one. Added the same comment.

Because the statement is already a full-width one-liner, the note goes on its own line immediately above rather than trailing, which keeps it inside the repo's line-length limits (`markdownlint`/`shfmt` clean, `shellcheck -x` clean):

```bash
  # not <<<: a >=64KiB here-string deadlocks (see hardcoded-path-patterns.sh)
  while IFS= read -r line || [[ -n "$line" ]]; do lines+=("$line"); done < <(printf '%s\n' "$body")
```

Your reasoning about *why* this file in particular matters is the load-bearing part, and it holds: `related_section` walks a PR body, and GitHub caps a body at exactly 65536 characters — inside the 65536–65663 hang window. A readability revert to `<<<"$body"` here would silently reintroduce a deadlock at the documented maximum body size.

Re-verified the function directly after the edit (a comment cannot change behavior, but this is a blocking gate, so I executed it rather than reasoned about it):

```text
good: Closes + Related                     -> PASS
good: No linked issue                      -> PASS
bad:  no Related section                   -> correctly reports missing section
bad:  empty Related                        -> correctly reports empty section
bad:  no keyword                           -> correctly reports missing keyword
nested ### is content                      -> PASS
final line w/o trailing newline retained   -> PASS
65600-byte body (deadlock window)          -> PASS in 1s
```

That last case is the one this whole PR exists for.

File: lib/hook-utils.sh [hunk 3]
```diff
@@ -406,12 +406,25 @@ hook::json_complete() {
   # the pre-filter can never turn a whole payload into a wrong verdict.
   [[ "${1: -4}" == *"}"* ]] || return 1
   command -v jq >/dev/null 2>&1 || return 1
-  # `printf | jq`, never `jq <<< "$1"`. A here-string is delivered through a pipe
-  # that bash fills itself, so a payload at or above the pipe capacity (65536
-  # bytes on this platform — exactly one read chunk) blocks the shell forever
-  # before jq is ever exec'd. Reproduced: a 65536-byte buffer hung here
-  # indefinitely while 65000 returned immediately. A separate writer process
-  # cannot deadlock that way.
+  # NEVER feed a whole payload through `<<<`. Bash delivers a here-string by
```
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply lib/hook-utils.sh]: Correct finding, and it was acted on — the diagnosis was exactly right.

`hook-utils-sync` did fail on the first push for precisely this reason. Reproduced locally at the time:

```text
$ scripts/sync-hook-utils.sh --check-bump origin/main
STALE VERSION: lib/hook-utils.sh changed vs origin/main but plugins/actionlint/... is still 0.8.0
… 14 stale manifests …
Bump the version of every carrying plugin so consumers receive the lib change.
rc=1
```

I took your second option — **removing the shared-helper edit** — in 78c3fd04, rather than bumping fourteen carriers. The edit was comment-only (tightening the here-string note's upper bound), and buying that inside a security fix at the cost of fourteen version bumps plus fourteen changelog entries was the wrong trade: it would have buried a two-guard fail-open fix in unrelated churn.

`lib/hook-utils.sh` is now byte-identical to `main`:

```text
$ git diff origin/main -- lib/hook-utils.sh | wc -c
0
$ scripts/sync-hook-utils.sh --check-bump origin/main
Lib unchanged vs origin/main; no version bumps required.   rc=0
```

`hook-utils-sync` and `hook-utils-windows` both pass on CI now.

Nothing was lost by dropping it. The contradiction this PR set out to reconcile lived on the **guardrails** side, where `hardcoded-path-patterns.sh` actively told readers to *prefer* a here-string over `printf | grep`. That advice is gone; the lib now states the same rule as `hook-utils.sh` and cites it. `hook-utils.sh`'s own guidance (`printf | jq`, never `jq <<<`) was already correct — only its upper bound was imprecise, which is a nicety, not a defect.

Resolving as fixed.

File: plugins/guardrails/CHANGELOG.md [hunk 4]
```diff
@@ -3,6 +3,74 @@
...
+  to `main`: its guidance was already correct, and the sync gate would require a version bump plus a
+  changelog entry for all fourteen other plugins that carry the shared lib in exchange for a
+  comment-only edit.
+
+- **The same deadlock in six command-scanning guards.** `block-convention-violation`,
+  `block-hook-bypass`, `flag-commit-pr-skill-bypass`, and the shared PowerShell command lib fed the
+  whole Bash/PowerShell command — or segments derived from it — through `while … done <<<"$cmd"`,
+  which deadlocks identically at 65536-65663 bytes. `workflow-resilience-check` did the same with an
+  inline Workflow `script:`. All now use `< <(printf '%s
+' …)`, which is byte-identical to the
```
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply plugins/guardrails/CHANGELOG.md]: Verified and fixed in 2895890c. Good catch, and the irony is sharp: this is the **same defect class** the PR fixes, from the same authoring mistake.

The eleven shell sites originally landed with a literal newline inside `printf '%s<LF>'` instead of `printf '%s\n'` — harmless there, since `printf` emits identical bytes either way, and corrected in 48d2e524. The same slip reached these CHANGELOG code spans, where it is **not** harmless: CommonMark folds a line ending inside a code span to a space, so the sentence explaining why `\n` is what makes the substitution byte-identical to the here-string rendered with the `\n` invisible.

Confirmed before fixing:

```text
$ grep -n "printf '%s$" plugins/guardrails/CHANGELOG.md
49:  inline Workflow `script:`. All now use `< <(printf '%s
```

Now a single line with the real two-character escape:

```text
$ grep -n "printf '%s" plugins/guardrails/CHANGELOG.md | cat -A
49:  inline Workflow `script:`. All now use `< <(printf '%s\n' …)`, which is byte-identical to the
```

`cat -A` shows `\n` as two literal characters with no `$` line-end between them, and the backticks on the line balance (4, even). `markdownlint-cli2`: 0 issues.

Both CHANGELOGs had it; the sibling thread on `source-control` is fixed in the same commit.

File: plugins/source-control/CHANGELOG.md [hunk 5]
```diff
@@ -3,6 +3,22 @@
...
+
+- **The PR-body linkage gate and validator deadlocked on a maximum-length PR body.** Both walked the
+  body with `while … done <<<"$body"`. Bash delivers a here-string by filling a pipe ITSELF, before
+  the reader is exec'd, and appends a newline, so a body of 65536-65663 bytes puts the write 1-128
+  bytes past the 65536-byte pipe capacity and blocks forever. GitHub caps a PR body at exactly
+  65536 characters, which lands INSIDE that window — so the worst case is not exotic, it is the
+  documented maximum. `pr-body-linkage-gate` is a blocking PreToolUse gate, so a hang means the
+  harness cancels it at its timeout and the linkage contract goes unenforced. Both now read through
+  `< <(printf '%s
+' "$body")`, which is byte-identical to the here-string it replaces and so cannot
```
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply plugins/source-control/CHANGELOG.md]: Verified and fixed in 2895890c, in the same commit as the `guardrails` sibling.

Confirmed before fixing:

```text
$ grep -n "printf '%s$" plugins/source-control/CHANGELOG.md
17:  `< <(printf '%s
```

Now one line carrying the real two-character escape:

```text
$ grep -n "printf '%s" plugins/source-control/CHANGELOG.md | cat -A
17:  `< <(printf '%s\n' "$body")`, which is byte-identical to the here-string it replaces and so
```

`cat -A` shows no `$` line-end between `%s` and `n`, backticks balance (2, even), and the reflowed wrap keeps the line inside the limit. `markdownlint-cli2`: 0 issues across both files.

Worth noting the sentence matters more here than anywhere else in the PR: this entry is the one explaining that GitHub's 65536-character PR-body cap lands inside the 65536–65663 hang window, and the `\n` is exactly what makes `< <(printf '%s\n' "$body")` byte-identical to the `<<<` it replaces — so no final line can be dropped from a body. Rendering that as `printf '%s '` would have undercut the claim in the sentence making it.
